#**Task 2: Classification Model & Feature Pipeline**

## Problem Objective

The objective of this task is to build a **classification model** that predicts whether a bank customer will accept a term deposit offer or not.

The Bank Marketing dataset contains customer and marketing information such as **age, job, education, marital status, loans, contact method, and previous campaign details**.

In this task, we will:

- Clean the dataset and handle missing or unknown values.
- Encode categorical features into numerical values.
- Normalize numerical features.
- Split the data into training and testing sets.
- Train a classification model to predict the customer's response.
- Evaluate the model using **accuracy, precision, recall, and F1-score**.

The target variable **`y`** represents the customer's response:

- `yes` → Customer accepted the term deposit offer.
- `no` → Customer did not accept the term deposit offer.

## Introduction to Bank Marketing

The **Bank Marketing dataset** is a dataset used for learning data analysis and machine learning.

The dataset contains information about **bank customers and marketing campaigns**. It is mainly used to predict whether a customer will accept a **term deposit offer**.

The dataset contains information about customers such as their **age, job, education, marital status, loans, contact information, and previous campaign results**.

## Target Variable

The main target variable in the dataset is **`y`**.

It tells us whether the customer accepted the term deposit offer:

- `yes` → Customer accepted the offer.
- `no` → Customer did not accept the offer.

This is the **target/label column** that our classification model will try to predict.

## Customer Features

The dataset contains several features that describe the customer.

### Age

- **`age`** represents the age of the customer.
- It is a numerical feature measured in years.

### Job

- **`job`** represents the type of job or occupation of the customer.
- Examples include `admin.`, `technician`, `services`, `blue-collar`, and `retired`.

### Marital Status

- **`marital`** represents the marital status of the customer.
- Examples include `married`, `single`, and `divorced`.

### Education

- **`education`** represents the education level of the customer.
- Examples include `high.school`, `university.degree`, and `basic.9y`.

## Financial Features

The dataset also contains information about the customer's financial situation.

- **`default`** indicates whether the customer has credit in default.
- **`housing`** indicates whether the customer has a housing loan.
- **`loan`** indicates whether the customer has a personal loan.

These features can help understand the customer's financial background.

## Contact and Campaign Features

The dataset contains information about how the customer was contacted and the marketing campaign.

- **`contact`** represents the communication method, such as telephone or cellular.
- **`month`** represents the month in which the customer was contacted.
- **`day_of_week`** represents the day of the week when the customer was contacted.
- **`campaign`** represents the number of contacts performed during the current campaign.
- **`previous`** represents the number of contacts performed before the current campaign.
- **`poutcome`** represents the outcome of the previous marketing campaign.

## Economic Features

The dataset also contains some economic indicators:

- **`emp.var.rate`** represents the employment variation rate.
- **`cons.price.idx`** represents the consumer price index.
- **`cons.conf.idx`** represents the consumer confidence index.
- **`euribor3m`** represents the 3-month Euribor interest rate.
- **`nr.employed`** represents the number of employees.

These features provide information about the economic conditions during the marketing campaign.

## Call Information

- **`duration`** represents the duration of the customer's last contact with the bank, measured in seconds.
- **`pdays`** represents the number of days since the customer was last contacted during a previous campaign.

For a realistic prediction model, `duration` can be removed because the call duration is only known after the customer has already been contacted.

## Features Used in the Dataset

The main features available for analysis are:

1. `age`
2. `job`
3. `marital`
4. `education`
5. `default`
6. `housing`
7. `loan`
8. `contact`
9. `month`
10. `day_of_week`
11. `duration`
12. `campaign`
13. `pdays`
14. `previous`
15. `poutcome`
16. `emp.var.rate`
17. `cons.price.idx`
18. `cons.conf.idx`
19. `euribor3m`
20. `nr.employed`

The **`y`** column is the **target/label column** because it tells us whether the customer accepted the term deposit offer.

There is no `Id` column in this dataset. Therefore, all columns other than the target `y` can potentially be used as features, although some features such as `duration` may be removed depending on the purpose of the model.

## Project Summary

This project focused on building a **classification model using the Bank Marketing dataset** to predict whether a customer would accept a term deposit offer.

The dataset contained **41,188 customer records and 21 columns**, including customer information, financial details, marketing campaign information, and economic indicators.

The data was first explored and cleaned by checking for missing values and `unknown` values. The features were then divided into **numerical and categorical features**. Numerical features were processed using **median imputation and StandardScaler**, while categorical features were processed using **most-frequent imputation and OneHotEncoder**.

A **Logistic Regression** model was trained using a complete preprocessing pipeline. The model achieved approximately **90% accuracy** on the test data.

However, the results showed that the model was much better at predicting customers who would **not accept** the offer than those who would **accept** it. The recall for the `yes` class was only **22%**, mainly because the dataset was imbalanced.

Overall, the project demonstrated the complete machine learning workflow, including **data cleaning, feature preprocessing, encoding, normalization, model training, prediction, and evaluation**. The model could be further improved by using techniques to handle class imbalance and by testing other classification algorithms.

#Step 1: Import Libraries

In [1]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

## Observation

The required Python libraries have been imported successfully for the **Bank Marketing classification task**.

- **Pandas** and **NumPy** will be used for data handling and numerical operations.
- **`train_test_split`** will be used to divide the dataset into training and testing sets.
- **`StandardScaler`** will normalize numerical features.
- **`OneHotEncoder`** will convert categorical features into numerical values.
- **`ColumnTransformer`** will apply different preprocessing steps to numerical and categorical features.
- **`Pipeline`** will combine preprocessing and the machine learning model into a single workflow.
- **`SimpleImputer`** will handle missing values.
- **`LogisticRegression`** will be used as the classification model.
- **`accuracy_score`**, **`classification_report`**, and **`confusion_matrix`** will be used to evaluate the model.

These libraries provide all the necessary tools to **clean, prepare, train, and evaluate** the Bank Marketing classification model.

#Step 2: Load the Dataset

In [2]:
df = pd.read_csv("/content/bank-additional-full.csv", sep=";")

#Display the first 5 rows

In [3]:
df.head()

,age,job,marital,education,default,housing,loan,contact,month,day_of_week,...,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y
0,56,housemaid,married,basic.4y,no,no,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
1,57,services,married,high.school,unknown,no,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
2,37,services,married,high.school,no,yes,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
3,40,admin.,married,basic.6y,no,no,no,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no
4,56,services,married,high.school,no,no,yes,telephone,may,mon,...,1,999,0,nonexistent,1.1,93.994,-36.4,4.857,5191.0,no


## Observation

The Bank Marketing dataset has been successfully loaded using Pandas.
- The dataset contains customer information such as `age`, `job`, `marital`, and `education`.
- It also contains marketing campaign information such as `campaign`, `pdays`, `previous`, and `poutcome`.
- Economic features such as `emp.var.rate`, `cons.price.idx`, and `euribor3m` are also included.
- The **`y`** column is the target variable, which indicates whether the customer accepted the term deposit offer (`yes` or `no`).
- Some categorical columns contain **`unknown`** values, which will need to be handled during data preprocessing.

The `head()` output shows the first **5 rows** of the dataset, giving us an initial understanding of its structure.


#Check the number of rows and columns

In [4]:
df.shape

(41188, 21)

## Observation

The dataset contains **41,188 rows and 21 columns**.

- **41,188 rows** represent the number of customer records in the dataset.
- **21 columns** represent the different customer, campaign, economic, and target variables.
- The dataset contains both **numerical and categorical features**.
- The target variable is **`y`**, which indicates whether the customer accepted the term deposit offer.

This shows that the dataset is sufficiently large for building and evaluating a classification model.

#Check column names

In [5]:
df.columns

Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'month', 'day_of_week', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'emp.var.rate', 'cons.price.idx',
       'cons.conf.idx', 'euribor3m', 'nr.employed', 'y'],
      dtype='object')

## Observation

The dataset contains **21 columns** representing different customer, marketing campaign, economic, and target information.

The columns can be broadly divided into:

- **Customer information:** `age`, `job`, `marital`, `education`
- **Financial information:** `default`, `housing`, `loan`
- **Contact information:** `contact`, `month`, `day_of_week`
- **Campaign information:** `duration`, `campaign`, `pdays`, `previous`, `poutcome`
- **Economic information:** `emp.var.rate`, `cons.price.idx`, `cons.conf.idx`, `euribor3m`, `nr.employed`
- **Target variable:** `y`

The **`y` column** is the target variable that we will predict using the other features.

#Check data types:

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null 

## Observation

The `info()` output shows that the dataset contains **41,188 records and 21 columns**.

- There are **5 integer columns** such as `age`, `duration`, `campaign`, `pdays`, and `previous`.
- There are **5 float columns** such as `emp.var.rate`, `cons.price.idx`, `cons.conf.idx`, `euribor3m`, and `nr.employed`.
- There are **11 object columns**, which contain categorical information such as `job`, `marital`, `education`, `contact`, and `y`.
- All columns show **41,188 non-null values**, meaning there are no actual `NaN` values in the dataset.
- However, some categorical columns contain **`unknown` values**, which will need to be handled as part of data cleaning.
- The **`y` column is an object type** because it contains the categorical values `yes` and `no`.

Therefore, the dataset contains both **numerical and categorical features**, so we will need different preprocessing techniques for each type.

#Step 4: Check Missing / Unknown Values

In [7]:
df.isnull().sum()

,0
age,0
job,0
marital,0
education,0
default,0
housing,0
loan,0
contact,0
month,0
day_of_week,0


## Observation

The `isnull().sum()` result shows that **all 21 columns have 0 missing (`NaN`) values**.

This means:

- There are **no actual null values** in the dataset.
- However, the dataset contains some **`unknown` values** in categorical columns such as `job`, `education`, and `default`.
- These `unknown` values are not detected by `isnull()`, because they are stored as text.
- Therefore, we still need to handle the `unknown` values during the data cleaning stage.

So, although the dataset has **no `NaN` values**, it still contains **unknown/missing information represented as `"unknown"`**.

In [8]:
(df == "unknown").sum()

,0
age,0
job,330
marital,80
education,1731
default,8597
housing,990
loan,990
contact,0
month,0
day_of_week,0


## Observation

The result shows that the dataset does not contain actual `NaN` values, but several categorical columns contain **`unknown`** values.

The columns with `unknown` values are:

- **`job`** → 330 unknown values
- **`marital`** → 80 unknown values
- **`education`** → 1,731 unknown values
- **`default`** → 8,597 unknown values
- **`housing`** → 990 unknown values
- **`loan`** → 990 unknown values

The `default` column has the **highest number of unknown values (8,597)**.

These `unknown` values represent missing or unavailable information and should be handled during data preprocessing. We can replace them with `NaN` and then use an imputation method to fill the missing categorical values.

#Step 5: Define Features and Target

In [9]:
X = df.drop("y", axis=1)
y = df["y"]

Here:
- X = input features
- y = target variable
- y = yes → customer accepted the offer
- y = no → customer did not accept the offer

#Step 6: Remove duration

In [10]:
X = X.drop("duration", axis=1)

##Why?
duration is the duration of the customer's call. You only know this after the call has taken place, so using it to predict the customer's response can cause data leakage.

### Why is this called data leakage?

Imagine we want to make a prediction **before calling the customer**:

```text
Customer information
        ↓
       MODEL
        ↓
Should we contact this customer?

At this point, we don't have duration.
But if we train the model with duration:

Customer information + duration
              ↓
            MODEL
              ↓
          Predict y

The problem is that duration is only known after the call has taken place. Therefore, the model is using information that would not be available at the time of the actual prediction.

This gives the model information from the future and can make the model's performance look better than it really is.

Therefore, using duration for a pre-call prediction can cause data leakage.

Data leakage occurs when information that would not be available at the time of prediction is used to train the model.

#Step 7: Split Numerical and Categorical Features

In [11]:
numerical_features = X.select_dtypes(include=["int64", "float64"]).columns

categorical_features = X.select_dtypes(include=["object"]).columns

print("Numerical Features:")
print(numerical_features)

print("\nCategorical Features:")
print(categorical_features)

Numerical Features:
Index(['age', 'campaign', 'pdays', 'previous', 'emp.var.rate',
       'cons.price.idx', 'cons.conf.idx', 'euribor3m', 'nr.employed'],
      dtype='object')

Categorical Features:
Index(['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact',
       'month', 'day_of_week', 'poutcome'],
      dtype='object')


## Observation

The features have been successfully divided into **numerical** and **categorical** features based on their data types.

### Numerical Features

There are **9 numerical features**:

- `age`
- `campaign`
- `pdays`
- `previous`
- `emp.var.rate`
- `cons.price.idx`
- `cons.conf.idx`
- `euribor3m`
- `nr.employed`

These features contain numerical values and will be **imputed and normalized using `StandardScaler`** during preprocessing.

### Categorical Features

There are **10 categorical features**:

- `job`
- `marital`
- `education`
- `default`
- `housing`
- `loan`
- `contact`
- `month`
- `day_of_week`
- `poutcome`

These features contain text or category values and will be **handled for missing/unknown values and encoded using `OneHotEncoder`**.

This separation is important because numerical and categorical features require **different preprocessing techniques** before training the classification model.

#Step 8: Split Training and Testing Data

In [12]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

This gives you:
- 80% training data
- 20% testing data
stratify=y helps maintain a similar proportion of yes and no responses in both sets.

#Step 9: Create the Preprocessing Pipeline

For numerical features:
- Missing values → Median → StandardScaler

For categorical features:
- Missing values → Most frequent → OneHotEncoder

In [13]:
numerical_pipeline = Pipeline([
    ("imputer", SimpleImputer(strategy="median")),
    ("scaler", StandardScaler())
])

categorical_pipeline = Pipeline([
    ("imputer", SimpleImputer(strategy="most_frequent")),
    ("encoder", OneHotEncoder(handle_unknown="ignore"))
])

In [14]:
preprocessor = ColumnTransformer([
    ("num", numerical_pipeline, numerical_features),
    ("cat", categorical_pipeline, categorical_features)
])



Since the dataset contains both numerical and categorical features, different preprocessing techniques are required for each type of feature.

### Numerical Features

For numerical features:

- Missing values are replaced with the **median** using `SimpleImputer`.
- Features are **standardized** using `StandardScaler`.

The process is:

**Missing Values → Median Imputation → Standardization**

### Categorical Features

For categorical features:

- Missing values are replaced with the **most frequent value**.
- Categorical values are converted into numerical form using **One-Hot Encoding**.

The process is:

**Missing Values → Most Frequent Value → One-Hot Encoding**



## Observation

The preprocessing pipeline has been created successfully.

- Numerical features will be filled using the **median** and then standardized.
- Categorical features will be filled using the **most frequent value** and then converted into numerical values using **One-Hot Encoding**.
- `ColumnTransformer` combines both preprocessing pipelines and applies the appropriate transformation to each group of features.

This prepares the dataset for the classification model.

#Step 10: Create Classification Model

In [15]:
model = Pipeline([
    ("preprocessor", preprocessor),
    ("classifier", LogisticRegression(max_iter=1000))
])

## Observation

A machine learning pipeline has been created by combining the **preprocessing steps** with a **Logistic Regression classification model**.

- The `preprocessor` applies the required transformations to numerical and categorical features.
- `LogisticRegression(max_iter=1000)` is used as the classification algorithm.
- The pipeline ensures that the same preprocessing steps are applied consistently to both training and testing data.
- This approach simplifies the workflow by combining **data preprocessing and model training** into a single pipeline.

The pipeline is now ready to be trained using the training dataset.

#Step 11: Train the Model

In [16]:
model.fit(X_train, y_train)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer(strategy='median')),
                                                                  ('scaler',
                                                                   StandardScaler())]),
                                                  Index(['age', 'campaign', 'pdays', 'previous', 'emp.var.rate',
       'cons.price.idx', 'cons.conf.idx', 'euribor3m', 'nr.employed'],
      dtype='object')),
                                                 ('cat',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer(strategy='most_frequent')),
                                                                  ('encoder',
                                                                   OneHotEncoder(handle_unknown='ignore'))]),
                                                  Index(['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact',
       'month', 'day_of_week', 'poutcome'],
      dtype='object'))])),
                ('classifier', LogisticRegression(max_iter=1000))])

## Observation

The classification model has been successfully trained using the training dataset.

The pipeline shows that:

- **Numerical features** are first processed using `SimpleImputer` and then normalized using `StandardScaler`.
- **Categorical features** are processed using `SimpleImputer` and then converted into numerical values using `OneHotEncoder`.
- The processed features are passed to the **Logistic Regression** classifier.
- All preprocessing steps and the classification model are combined into a single pipeline.

Therefore, the model is now ready to make predictions on the **test dataset** and evaluate its performance.

#Step 12: Make Predictions

In [17]:
y_pred = model.predict(X_test)

#Step 13: Evaluate the Model

In [18]:
accuracy = accuracy_score(y_test, y_pred)

print("Accuracy:", accuracy)

Accuracy: 0.9009468317552805


## Observation

The trained Logistic Regression model was used to make predictions on the test dataset using `model.predict(X_test)`.

The model achieved an **accuracy of approximately 90.05%**.

This means that the model correctly predicted the customer's response for about **90 out of every 100 test records**.

The accuracy indicates that the model has performed well overall. However, since the dataset contains more `no` responses than `yes` responses, accuracy alone may not provide a complete picture of the model's performance. Therefore, **precision, recall, F1-score, and the confusion matrix** should also be checked.

#Classification Report

In [19]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

          no       0.91      0.99      0.95      7310
         yes       0.69      0.22      0.33       928

    accuracy                           0.90      8238
   macro avg       0.80      0.60      0.64      8238
weighted avg       0.88      0.90      0.88      8238



## Observation

The classification report shows how well the model predicts the two possible results: `no` and `yes`.

- The model predicts **`no` customers very well**, with a precision of **91%** and recall of **99%**.
- For **`yes` customers**, the model is less effective. Its precision is **69%**, but its recall is only **22%**.
- This means the model correctly identifies only a small number of customers who would actually accept the term deposit.
- There are **7,310 `no` cases** and only **928 `yes` cases**, so the dataset is **imbalanced**.
- The model has an overall **accuracy of 90%**, which looks good.
- However, accuracy alone is not enough because the model is much better at predicting `no` than `yes`.

**In simple words:** The model is good at identifying customers who will **not** accept the offer, but it needs improvement in identifying customers who **will** accept the offer.

#Confusion Matrix

In [20]:
print(confusion_matrix(y_test, y_pred))

[[7219   91]
 [ 725  203]]


It shows how correctly and incorrectly the model predicted the two classes.
- 7,219 customers were correctly predicted as no.
- 203 customers were correctly predicted as yes.
- 91 customers were actually no, but the model predicted yes.
- 725 customers were actually yes, but the model predicted no.
The most important point is that the model correctly identified only 203 out of 928 yes customers. It missed 725 customers who would actually accept the term deposit.

In simple words: The model is very good at predicting customers who will say no, but it misses many customers who would say yes. This is why the recall for the yes class is low.

## Key Insights

- The Bank Marketing dataset contains **41,188 customer records and 21 columns**.
- The dataset includes both **numerical and categorical features**, so different preprocessing methods were required.
- There were no actual `NaN` values, but several columns contained **`unknown` values** that needed to be handled.
- The target variable `y` contains two classes: **`yes` and `no`**.
- The dataset is **imbalanced**, with significantly more `no` responses than `yes` responses.
- A preprocessing pipeline was created using **median imputation and StandardScaler** for numerical features.
- Categorical features were processed using **most-frequent imputation and OneHotEncoder**.
- A **Logistic Regression** model was trained using the complete preprocessing pipeline.
- The model achieved an overall **accuracy of approximately 90%**.
- The model performed very well for predicting `no`, with a **99% recall**.
- The model performed poorly at identifying `yes` customers, with only **22% recall**.
- The confusion matrix showed that **725 actual `yes` customers were incorrectly predicted as `no`**.
- Therefore, the main limitation of the model is its **low ability to identify customers who are likely to accept the term deposit**.
- Future improvement could include handling the class imbalance using techniques such as **class weights, oversampling, or SMOTE**, and comparing other classification models.